### PR TITLE
Add tests to language part 1

### PR DESCRIPTION
**Pull request by bingbing8**
Add 5 test files and 1 help files for language.

### CODE REVIEW

File: test/powershell/Language/Scripting/ActionPreference.Tests.ps1 [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        }
+        
+        It 'ActionPreference Ignore Works' {            
+            $errorCount = $error.Count
+            Get-Process -Name asdfasdfsadfsadf -ErrorAction Ignore            
+
+            $error.Count | Should Be $errorCount
+        }
+        
+        Context 'action preference of Ignore cannot be set as a preference variable' {
```
Comment (JamesWTruher) [inline test/powershell/Language/Scripting/ActionPreference.Tests.ps1]: >Context [](start = 8, length = 7)

this should probably be `It`

File: test/powershell/Language/Scripting/ActionPreference.Tests.ps1 [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+            try { 
+                    $GLOBAL:errorActionPreference = "Suspend"
+                    Get-Process -Name asdfasdfasdf
+                    Throw "Execution OK"
+                } catch { 
+                    $_.CategoryInfo.Reason | Should Be ArgumentTransformationMetadataException
+                }
+        }
+
+        It 'enum disambiguation works' {
```
Comment (JamesWTruher) [inline test/powershell/Language/Scripting/ActionPreference.Tests.ps1]: >t 'enum disambiguation works' [](start = 9, length = 29)

you've been changing the global state of errorActionPreference above - could that cause problems in this test?

File: test/powershell/Language/Scripting/TestsOnWinFullOnly.ps1 [hunk 3]
```diff
@@ -0,0 +1,27 @@
+﻿function Run-TestOnWinFull
+{
+    [CmdletBinding()]
+    param( [string]$name )
+
+    switch ($name)
```
Comment (JamesWTruher) [inline test/powershell/Language/Scripting/TestsOnWinFullOnly.ps1]: should this have a default case?

File: test/powershell/Language/Scripting/ActionPreference.Tests.ps1 [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+            {
+                Get-Process -ErrorAction Suspend
+                Throw "Execution OK"
+            }
+            catch {
+                $_.FullyQualifiedErrorId | Should Be "ParameterBindingFailed,Microsoft.PowerShell.Commands.GetProcessCommand"
+            }
+        }
+
+        Context 'WarningAction = Suspend does not work' {
```
Comment (JamesWTruher) [inline test/powershell/Language/Scripting/ActionPreference.Tests.ps1]: >Context [](start = 8, length = 7)

should this be `It`?

File: test/powershell/Language/Scripting/ActionPreference.Tests.ps1 [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+            }
+            catch {
+                $_.FullyQualifiedErrorId | Should Be "ParameterBindingFailed,Microsoft.PowerShell.Commands.GetProcessCommand"
+            }
+        }
+
+        #issue 2076
+        It 'ErrorAction and WarningAction are the only action preferences do not support suspend' -Pending{
+
+            $params = [System.Management.Automation.Internal.CommonParameters].GetProperties().Name | sls Action
```
Comment (JamesWTruher) [inline test/powershell/Language/Scripting/ActionPreference.Tests.ps1]: >sls [](start = 102, length = 3)

please don't use aliases

File: test/powershell/Language/Scripting/CheckRestrictedlanguage.Tests.ps1 [hunk 6]
```diff
@@ -0,0 +1,112 @@
+﻿#  <Test>
+#    <summary>Test restricted language check method on scriptblocks</summary>
+#  </Test>
+Describe "Test restricted language check method on scriptblocks" -Tags "CI" {
+        BeforeAll {
+            set-strictmode -v 2
+            function list {
+
+            $l = new-object system.collections.generic.list[string]
```
Comment (JamesWTruher) [inline test/powershell/Language/Scripting/CheckRestrictedlanguage.Tests.ps1]: >w-object system.collections.generic.list[string] [](start = 19, length = 48)

i think the form we're suggesting now is:
`[System.Collections.Generic.List[String]]::new()`

File: test/powershell/Language/Scripting/CheckRestrictedlanguage.Tests.ps1 [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+                { $a + $b + $c }.CheckRestrictedLanguage($null, (list a b), $false)   # fail
+                Throw "Execution OK"
+            }
+            catch
+            {
+                $_.FullyQualifiedErrorId | Should Be 'ParseException'
+            }
+        }
+
+        It 'Check union of default + one allowed variable and but not allow evironment variable' {
```
Comment (JamesWTruher) [inline test/powershell/Language/Scripting/CheckRestrictedlanguage.Tests.ps1]: >evironment [](start = 76, length = 10)

environment

File: test/powershell/Language/Scripting/CheckRestrictedlanguage.Tests.ps1 [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+                { 2 + $a + $env:foo }.CheckRestrictedLanguage($null, (list a), $false)   # fail
+                Throw "Execution OK"
+            }
+            catch
+            {
+                $_.FullyQualifiedErrorId | Should Be 'ParseException'
+            }
+        }
+
+        It 'Check union of default + one allowed varialbe name and allow environment variable ' {
```
Comment (JamesWTruher) [inline test/powershell/Language/Scripting/CheckRestrictedlanguage.Tests.ps1]: >varialbe [](start = 49, length = 8)

variable

File: test/powershell/Language/Scripting/CmdletDeclaration.Tests.ps1 [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+                    param()
+
+	            function get-bar
+                    {
+                        param([Parameter(mandatory=$true)]$a)
+                        $a
+                    }
+                }'}
+                @{
+                Name = 'cmdlet declaration within scripblock';
```
Comment (JamesWTruher) [inline test/powershell/Language/Scripting/CmdletDeclaration.Tests.ps1]: >scripblock [](start = 50, length = 10)

scriptblock

File: test/powershell/Language/Scripting/CmdletDeclaration.Tests.ps1 [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+                Script = '
+                {
+                    param([Parameter(mandatory=$true)]$a)
+                    $a
+                }'},
+                @{
+                Name = 'cmdlet declaration within scripblock 2';
```
Comment (JamesWTruher) [inline test/powershell/Language/Scripting/CmdletDeclaration.Tests.ps1]: >scripblock [](start = 50, length = 10)

scriptblock

File: test/powershell/Language/Scripting/CmdletDeclaration.Tests.ps1 [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+                Script = '
+                {
+                    param([Parameter(mandatory=$true)]$a)
+                    $a
+                }'},
+                @{
+                Name = 'cmdlet declaration within scripblock 3';
```
Comment (JamesWTruher) [inline test/powershell/Language/Scripting/CmdletDeclaration.Tests.ps1]: >scripblock [](start = 50, length = 10)

scriptblock

File: test/powershell/Language/Scripting/ActionPreference.Tests.ps1 (hunk 2 above)
Comment (bingbing8) [inline test/powershell/Language/Scripting/ActionPreference.Tests.ps1]: Please see line 8. I reset it to original value. Also this line will not able to change the value.

File: test/powershell/Language/Scripting/ActionPreference.Tests.ps1 (hunk 1 above)
Comment (bingbing8) [inline test/powershell/Language/Scripting/ActionPreference.Tests.ps1]: fixed.

File: test/powershell/Language/Scripting/TestsOnWinFullOnly.ps1 (hunk 3 above)
Comment (bingbing8) [inline test/powershell/Language/Scripting/TestsOnWinFullOnly.ps1]: I will add a default, but do nothing

File: test/powershell/Language/Scripting/ActionPreference.Tests.ps1 (hunk 4 above)
Comment (bingbing8) [inline test/powershell/Language/Scripting/ActionPreference.Tests.ps1]: fixed.

File: test/powershell/Language/Scripting/ActionPreference.Tests.ps1 (hunk 5 above)
Comment (bingbing8) [inline test/powershell/Language/Scripting/ActionPreference.Tests.ps1]: fixed.

File: test/powershell/Language/Scripting/CheckRestrictedlanguage.Tests.ps1 (hunk 6 above)
Comment (bingbing8) [inline test/powershell/Language/Scripting/CheckRestrictedlanguage.Tests.ps1]: ok.

File: test/powershell/Language/Scripting/CheckRestrictedlanguage.Tests.ps1 (hunk 7 above)
Comment (bingbing8) [inline test/powershell/Language/Scripting/CheckRestrictedlanguage.Tests.ps1]: fixed

File: test/powershell/Language/Scripting/CheckRestrictedlanguage.Tests.ps1 (hunk 8 above)
Comment (bingbing8) [inline test/powershell/Language/Scripting/CheckRestrictedlanguage.Tests.ps1]: fixed

File: test/powershell/Language/Scripting/CmdletDeclaration.Tests.ps1 (hunk 9 above)
Comment (bingbing8) [inline test/powershell/Language/Scripting/CmdletDeclaration.Tests.ps1]: fixed

File: test/powershell/Language/Scripting/CmdletDeclaration.Tests.ps1 (hunk 10 above)
Comment (bingbing8) [inline test/powershell/Language/Scripting/CmdletDeclaration.Tests.ps1]: fixed

File: test/powershell/Language/Scripting/CmdletDeclaration.Tests.ps1 (hunk 11 above)
Comment (bingbing8) [inline test/powershell/Language/Scripting/CmdletDeclaration.Tests.ps1]: fixed

File: test/powershell/Language/Scripting/TestsOnWinFullOnly.ps1 (hunk 3 above)
Comment (bingbing8) [inline test/powershell/Language/Scripting/TestsOnWinFullOnly.ps1]: I added a default, but do nothing

File: test/powershell/Language/Scripting/ActionPreference.Tests.ps1 (hunk 2 above)
Comment (bingbing8) [inline test/powershell/Language/Scripting/ActionPreference.Tests.ps1]: Added a finally block to reset the global variable.

File: test/powershell/Language/Scripting/ActionPreference.Tests.ps1 (hunk 1 above)
Comment (bingbing8) [inline test/powershell/Language/Scripting/ActionPreference.Tests.ps1]: fixed

---
In reply to: [76494516](https://github.com/powershell/powershell/pull/2086#discussion_r76494516) [](ancestors = 76494516)

File: test/powershell/Language/Scripting/ActionPreference.Tests.ps1 (hunk 2 above)
Comment (bingbing8) [inline test/powershell/Language/Scripting/ActionPreference.Tests.ps1]: Added a finally block to reset the global variable.

---
In reply to: [76494701](https://github.com/powershell/powershell/pull/2086#discussion_r76494701) [](ancestors = 76494701)

File: test/powershell/Language/Scripting/ActionPreference.Tests.ps1 (hunk 4 above)
Comment (bingbing8) [inline test/powershell/Language/Scripting/ActionPreference.Tests.ps1]: fixed

---
In reply to: [76494914](https://github.com/powershell/powershell/pull/2086#discussion_r76494914) [](ancestors = 76494914)

File: test/powershell/Language/Scripting/ActionPreference.Tests.ps1 (hunk 5 above)
Comment (bingbing8) [inline test/powershell/Language/Scripting/ActionPreference.Tests.ps1]: fixed

---
In reply to: [76494951](https://github.com/powershell/powershell/pull/2086#discussion_r76494951) [](ancestors = 76494951)

File: test/powershell/Language/Scripting/CheckRestrictedlanguage.Tests.ps1 (hunk 6 above)
Comment (bingbing8) [inline test/powershell/Language/Scripting/CheckRestrictedlanguage.Tests.ps1]: fixed

---
In reply to: [76495042](https://github.com/powershell/powershell/pull/2086#discussion_r76495042) [](ancestors = 76495042)

File: test/powershell/Language/Scripting/CheckRestrictedlanguage.Tests.ps1 (hunk 7 above)
Comment (bingbing8) [inline test/powershell/Language/Scripting/CheckRestrictedlanguage.Tests.ps1]: fixed

---
In reply to: [76495139](https://github.com/powershell/powershell/pull/2086#discussion_r76495139) [](ancestors = 76495139)

File: test/powershell/Language/Scripting/CheckRestrictedlanguage.Tests.ps1 (hunk 8 above)
Comment (bingbing8) [inline test/powershell/Language/Scripting/CheckRestrictedlanguage.Tests.ps1]: fixed

---
In reply to: [76495185](https://github.com/powershell/powershell/pull/2086#discussion_r76495185) [](ancestors = 76495185)

File: test/powershell/Language/Scripting/CmdletDeclaration.Tests.ps1 (hunk 9 above)
Comment (bingbing8) [inline test/powershell/Language/Scripting/CmdletDeclaration.Tests.ps1]: fixed

---
In reply to: [76495751](https://github.com/powershell/powershell/pull/2086#discussion_r76495751) [](ancestors = 76495751)

File: test/powershell/Language/Scripting/CmdletDeclaration.Tests.ps1 (hunk 10 above)
Comment (bingbing8) [inline test/powershell/Language/Scripting/CmdletDeclaration.Tests.ps1]: fixed

---
In reply to: [76495765](https://github.com/powershell/powershell/pull/2086#discussion_r76495765) [](ancestors = 76495765)

File: test/powershell/Language/Scripting/CmdletDeclaration.Tests.ps1 (hunk 11 above)
Comment (bingbing8) [inline test/powershell/Language/Scripting/CmdletDeclaration.Tests.ps1]: fixed

---
In reply to: [76495785](https://github.com/powershell/powershell/pull/2086#discussion_r76495785) [](ancestors = 76495785)

File: test/powershell/Language/Scripting/TestsOnWinFullOnly.ps1 (hunk 3 above)
Comment (bingbing8) [inline test/powershell/Language/Scripting/TestsOnWinFullOnly.ps1]: fixed

---
In reply to: [76494845](https://github.com/powershell/powershell/pull/2086#discussion_r76494845) [](ancestors = 76494845)

**Comment by JamesWTruher** [comment]
:shipit:

File: test/powershell/Language/Scripting/ActionPreference.Tests.ps1 (hunk 2 above)
Comment (bingbing8) [inline test/powershell/Language/Scripting/ActionPreference.Tests.ps1]: Can you review? the finally block was added in previous tests to reset the global variable.

---
In reply to: [76505667](https://github.com/Powershell/Powershell/pull/2086#discussion_r76505667) [](ancestors = 76505667,76494701)

**Comment by vors** [comment]
@bingbing8 the build is still failing. Did you have a chance to investigate it?

**Comment by bingbing8** [comment]
I am still investigating.

---
In reply to: [245731706](https://github.com/Powershell/Powershell/pull/2086#issuecomment-245731706) [](ancestors = 245731706)

**Comment by bingbing8** [comment]
@vors, I found the issue and got it fixed. Please check.

**Comment by bingbing8** [comment]
@vors  finally found the issue and got it fixed. please check.

**Comment by vors** [comment]
Great! Now, we need to make the commits history better, so we keep a high-quality bar for the repo history.

Particularly: 

* Currently there are 29 commits. It's unreasonable big number for PR that adds 7 new test files.
* History contains your experiments to troubleshoot the failure on CI. That should not be included in the final history.
* History contains two commits with a message "fixing indentation". At least, it's redundant to have two of them. I would question even having one of them, because it's probably better to incorporate in the previous commits, since this is a new code.
* Commit messages don't follow [contribution guidelines](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#common-engineering-practices). Particularly in f4041463a9ceadb16043e02ba4eb9c87759120ce it's desirable to add more details to the commit message body about tests that are added.

Now, there are two approaches we can take here:
1. I can just squash all changes in the merge UI and we will have 1 commit for everything. It could be acceptable **in this particular** case.
2. You can do an [intercative rebase](https://git-scm.com/docs/git-rebase) to re-write history. It will also be very useful for your ongoing contributions because these git history questions will come again and again.

Please, review the [pull request lifecycle](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#lifecycle-of-a-pull-request) for more details.

**Pull request merged**